cloudinit/sources/DataSourceAzure.py

```diff
@@ -634,6 +634,48 @@ def _get_data(self):
             self.ds_cfg['data_dir'], crawled_data['files'], dirmode=0o700)
         return True
 
+    @azure_ds_telemetry_reporter
+    def get_imds_data(
+            self,
+            fallback_nic,
+            retries,
+            md_type=metadata_type.compute):
+        LOG.info("Attempting IMDS api-version: %s", self.api_version)
+        if self.failed_desired_api_version:
+            return get_metadata_from_imds(
+                fallback_nic=fallback_nic,
+                retries=retries,
+                md_type=md_type,
+                api_version=self.api_version
+            )
+
+        try:
+            try:
+                return get_metadata_from_imds(
+                    fallback_nic=fallback_nic,
+                    retries=1,
+                    md_type=md_type,
+                    api_version=self.api_version
+                )
+            except UrlError as err:
+                LOG.info(
+                    "UrlError with IMDS api-version: %s",
+                    self.api_version
+                )
+                if err.code == 400:
+                    self.api_version = IMDS_VER_MIN
+                    log_msg = "Falling back to IMDS api-version: {}".format(
+                        self.api_version
+                    )
+                    report_diagnostic_event(
+                        log_msg,
+                        logger_func=LOG.info
+                    )
+                    self.failed_desired_api_version = True
+                raise
+        except Exception:
+            return self.get_imds_data(fallback_nic, retries - 1, md_type)
```

Comment:
I think a non-recursive solution might be more straightforward. Would something like this be equivalent to what's happening in this method?
```python
        if not self.failed_desired_api_version:
            for _ in range(retries):
                try:
                    return get_metadata_from_imds(
                        fallback_nic=fallback_nic,
                        retries=0,
                        md_type=md_type,
                        api_version=IMDS_VER_WANT,
                    )
                except UrlError as err:
                    # Log stuff...
                    if err.code == 400:
                        # Log stuff and report diagnostic event
                        self.failed_desired_api_version = True
                        break
        return get_metadata_from_imds(
            fallback_nic=fallback_nic,
            retries=retries,
            md_type=md_type,
            api_version=IMDS_VER_MIN,
        )
```
This would also remove the need for a separate `api_version` instance variable (assuming the test could live without it).